homeassistant/components/xiaomi_miio/switch.py

```diff
@@ -113,120 +87,111 @@
+        elif model in [
+            "chuangmi.plug.m1",
+            "chuangmi.plug.m3",
+            "chuangmi.plug.v2",
+            "chuangmi.plug.hmi205",
+            "chuangmi.plug.hmi206",
+        ]:
+            plug = ChuangmiPlug(host, token, model=model)
+            device = XiaomiPlugGenericSwitch(name, plug, config_entry, unique_id)
+            entities.append(device)
+            hass.data[DATA_KEY][host] = device
+        elif model in ["lumi.acpartner.v3"]:
+            plug = AirConditioningCompanionV3(host, token)
+            device = XiaomiAirConditioningCompanionSwitch(
+                name, plug, config_entry, unique_id
+            )
+            entities.append(device)
+            hass.data[DATA_KEY][host] = device
+        else:
+            _LOGGER.error(
+                "Unsupported device found! Please create an issue at "
+                "https://github.com/rytilahti/python-miio/issues "
+                "and provide the following data: %s",
                 model,
-                device_info.firmware_version,
-                device_info.hardware_version,
             )
-        except DeviceException as ex:
-            raise PlatformNotReady from ex
-
-    if model in ["chuangmi.plug.v1", "chuangmi.plug.v3", "chuangmi.plug.hmi208"]:
-        plug = ChuangmiPlug(host, token, model=model)
 
-        # The device has two switchable channels (mains and a USB port).
-        # A switch device per channel will be created.
-        for channel_usb in [True, False]:
-            device = ChuangMiPlugSwitch(name, plug, model, unique_id, channel_usb)
-            devices.append(device)
-            hass.data[DATA_KEY][host] = device
+        async def async_service_handler(service):
+            """Map services to methods on XiaomiPlugGenericSwitch."""
+            method = SERVICE_TO_METHOD.get(service.service)
+            params = {
+                key: value
+                for key, value in service.data.items()
+                if key != ATTR_ENTITY_ID
+            }
+            entity_ids = service.data.get(ATTR_ENTITY_ID)
+            if entity_ids:
+                devices = [
+                    device
+                    for device in hass.data[DATA_KEY].values()
+                    if device.entity_id in entity_ids
+                ]
+            else:
+                devices = hass.data[DATA_KEY].values()
+
+            update_tasks = []
+            for device in devices:
+                if not hasattr(device, method["method"]):
+                    continue
+                await getattr(device, method["method"])(**params)
+                update_tasks.append(device.async_update_ha_state(True))
+
+            if update_tasks:
+                await asyncio.wait(update_tasks)
+
+        for plug_service in SERVICE_TO_METHOD:
+            schema = SERVICE_TO_METHOD[plug_service].get("schema", SERVICE_SCHEMA)
+            hass.services.async_register(
```

Comment:
If we use our entity service helper a lot of this code can be cleaned up.

https://developers.home-assistant.io/docs/dev_101_services#entity-services

Reply:
I had also noticed that and I completly agree that this schould be done.
However that is out of scope for this PR.

I want to implement config flow first for all Miio platforms (also for vacuum, light, device_tracker etc etc.).
Another thing that needs to be done is move all the state_attributes to seperate binary_sensors/sensors.
This config flow is the first step for that.

But I don't want to make these PR's to big, therfore I want to leave the original code as much as possible and only change from yaml.config to config flow.

Follow up PRs will take care of the attributes and the services.